qutebrowser/misc/sql.py

```diff
@@ -0,0 +1,232 @@
+        q.run()
+        return iter(q)
+
+    def contains_query(self, field):
+        """Return a prepared query that checks for the existence of an item.
+
+        Args:
+            field: Field to match.
+        """
+        return Query("Select EXISTS(SELECT * FROM {} where {} = ?)"
+                     .format(self._name, field))
+
+    def __len__(self):
+        """Return the count of rows in the table."""
+        q = Query("SELECT count(*) FROM {}".format(self._name))
+        q.run()
+        return q.value()
+
+    def delete(self, value, field):
+        """Remove all rows for which `field` equals `value`.
+
+        Args:
+            value: Key value to delete.
+            field: Field to use as the key.
+
+        Return:
+            The number of rows deleted.
+        """
+        q = Query("DELETE FROM {} where {} = ?".format(self._name, field))
+        q.run([value])
+        if not q.numRowsAffected():
+            raise KeyError('No row with {} = "{}"'.format(field, value))
+        self.changed.emit()
+
+    def insert(self, values, replace=False):
+        """Append a row to the table.
+
+        Args:
+            values: A list of values to insert.
+            replace: If set, replace existing values.
+        """
+        paramstr = ','.join(['?'] * len(values))
+        q = Query("INSERT {} INTO {} values({})".format(
+            'OR REPLACE' if replace else '', self._name, paramstr))
+        q.run(values)
+        self.changed.emit()
+
+    def insert_batch(self, rows, replace=False):
+        """Performantly append multiple rows to the table.
+
+        Args:
+            rows: A list of lists, where each sub-list is a row.
+            replace: If set, replace existing values.
+        """
+        paramstr = ','.join(['?'] * len(rows[0]))
+        q = Query("INSERT {} INTO {} values({})".format(
+            'OR REPLACE' if replace else '', self._name, paramstr))
+
+        transposed = [list(row) for row in zip(*rows)]
+        for val in transposed:
+            q.addBindValue(val)
+
+        db = QSqlDatabase.database()
+        db.transaction()
+        if not q.execBatch():
+            raise SqlException('Failed to exec query "{}": "{}"'.format(
+                               q.lastQuery(), q.lastError().text()))
+        db.commit()
+        self.changed.emit()
+
+    def delete_all(self):
+        """Remove all row from the table."""
+        Query("DELETE FROM {}".format(self._name)).run()
+        self.changed.emit()
+
+    def select(self, sort_by, sort_order, limit=-1):
+        """Remove all row from the table.
+
+        Args:
+            sort_by: name of column to sort by.
+            sort_order: 'asc' or 'desc'.
+            limit: max number of rows in result, defaults to -1 (unlimited).
+        """
+        q = Query('SELECT * FROM {} ORDER BY {} {} LIMIT ?'
+                  .format(self._name, sort_by, sort_order))
+        q.run([limit])
```

Comment:
I've been using placeholders wherever possible just because I figured it was safer.